flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java

```diff
@@ -212,6 +212,16 @@ protected void channelRead0(final ChannelHandlerContext ctx, final HttpObject ms
         }
     }
 
+    private static boolean hasNext(HttpPostRequestDecoder decoder) {
+        try {
+            return decoder.hasNext();
+        } catch (HttpPostRequestDecoder.EndOfDataDecoderException e) {
+            // this can occur if the final chuck wasn't empty, but didn't contain any attribute data
+            // unfortunately the Netty APIs don't give us any way to check this
```

Comment:
we could maintain the size of `currentHttpPostRequestDecoder.getBodyHttpDatas()` (that's essentially what the `hasNext()`/`next()` methods access and which increases with every `offer(..)` call): Only if the size of the returned List increases we now that there is more data to read. 

That's the only workaround I could come up with based on what I found from reading through the `HttpPostMultipartRequestDecoder` iteration code and from what I found in https://github.com/netty/netty/issues/7869. Not sure whether that's a nicer approach, though. :thinking:

Reply:
> we could maintain the size of currentHttpPostRequestDecoder.getBodyHttpDatas() 

You can only call `getBodyHttpDatas` once you've read everything.

Reply:
Which sucks a bit since you can't fail early on unknown attributes for example, and it'd be a larger change.

Reply:
ah true, I missed the part again where the `NotEnoughDataDecoderException` is thrown :facepalm:

Reply:
> ah true, I missed the part again where the NotEnoughDataDecoderException is thrown 🤦

Found that out the hard way :sob:

Reply:
the contract of hasNext method seems just awful (I'm still puzzled they simply don't return false), but it seems that netty even recommends this approach in examples

https://github.com/netty/netty/blob/6fc60e04649ae3873f0f52ecb3a387245215dbad/example/src/main/java/io/netty/example/http/upload/HttpUploadServerHandler.java#L252

Reply:
interesting finding :+1: